[TensorPipe] Use the new multi-payload message API

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37919 [TensorPipe] Use the new multi-payload message API**
* #37918 [TensorPipe] Allow passing args to agent options constructor

In D21209901 TensorPipe added support for a vector of payloads inside each message, instead of a single one, so that users with multiple payloads can send them separately as they are instead of having to copy them into a new block of contiguous memory. The PyTorch agent is using the old API, which is preventing us from deleting it. This change has no effects on over-the-wire format and thus on performance.



Differential Revision: [D21425536](https://our.internmc.facebook.com/intern/diff/D21425536/)

Differential Revision: [D21425536](https://our.internmc.facebook.com/intern/diff/D21425536)

lw committed:
[TensorPipe] Use the new multi-payload message API



Differential Revision: [D21425536](https://our.internmc.facebook.com/intern/diff/D21425536/)

[ghstack-poisoned]

lw committed:
Update on "[TensorPipe] Use the new multi-payload message API"


Differential Revision: [D21425536](https://our.internmc.facebook.com/intern/diff/D21425536/)

[ghstack-poisoned]

mrshenli commented on test/cpp/rpc/test_tensorpipe_serialization.cpp:
```diff
@@ -25,12 +25,20 @@ TEST(TensorpipeSerialize, Base) {
   torch::distributed::rpc::TensorPipeEntry tpEntry =
       torch::distributed::rpc::tensorpipeSerialize(sendingRpcMessage);
   tensorpipe::Message sendingTpMessage = std::move(tpEntry.message);
-  EXPECT_EQ(sendingTpMessage.tensors.size(), 2);
```
Why removing this check? Is the number of tensors no longer 2?

lw replied:
I think the main purpose of this test is that `tensorpipeAllocateMessage` is the inverse of `tensorpipeSerialize`, that is, that by passing through one and then the other one obtains the original RPC message. The implementation details of the format that these two functions choose to adopt is irrelevant to the correctness of the test, as long as they both agree on it. So for example even if the RPC message contained two tensors this doesn't have to be the case for the TP message: it could deduplicate them (if they have the same storage), coalesce them, treat one as a payload (if it's small enough that latency is more important than bandwidth), ... The test should be robust to these changes.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -122,15 +122,6 @@ void TensorPipeAgent::pipeRead(
     // Allocate memory and fill in pointers
     Message rpcMessage = tensorpipeAllocateMessage(tpMessage);
-    TORCH_INTERNAL_ASSERT(
-        rpcMessage.tensors().size() == tpMessage.tensors.size(),
```
Looks like `tensorpipeAllocateMessage` does add tensors into `rpcMessage`, why removing this check here?

lw replied:
Same reason as above, the number of tensors in the TP message is an implementation detail and may not be equal to the number of tensors in the RPC message.

jiayisuse replied:
Right now tensorpipe id not deduplicating tensors sharing the same storage. I think the number of tensors between RPC message and TensorPipe message are always the same. Let's keep these checks until handling duplicated tensor storage?

lw replied:
Yes, it is true that at the moment that condition would be true and the check would pass. My take is that tests should only be allowed to check what is in the specification of a function, i.e., the semantics of its interface. If they do instead check the implementation then they may need to be updated each time the code changes which means they are not robust and don't really provide any safeguard.

jiayisuse replied:
This guard is right to make sure each update of the implementation doesn't break the critical assumption, unless we start to handle the duplicated tensor storage, which will break the assumption.

lw replied:
The only "critical" assumption, IMO, is that deserializing a serialized message gives back the original. Which is now exactly the only thing the tests are checking.

mrshenli reviewed (commented):
Could you please add some description to this PR?

lw commented:
Sorry, I added a descriptor to the diff after I had exported it, let me re-export it.

lw committed:
Update on "[TensorPipe] Use the new multi-payload message API"


In D21209901 TensorPipe added support for a vector of payloads inside each message, instead of a single one, so that users with multiple payloads can send them separately as they are instead of having to copy them into a new block of contiguous memory. The PyTorch agent is using the old API, which is preventing us from deleting it. This change has no effects on over-the-wire format and thus on performance.



Differential Revision: [D21425536](https://our.internmc.facebook.com/intern/diff/D21425536/)

Differential Revision: [D21425536](https://our.internmc.facebook.com/intern/diff/D21425536)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Test failures are irrelevant.

jiayisuse commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -412,9 +414,15 @@ TensorPipeEntry tensorpipeSerialize(const Message& rpcMessage) {
+Message tensorpipeAllocateMessage(tensorpipe::Message& tpMessage) {
   // Payload, message type and message id
-  std::vector<char> payload(tpMessage.length);
+  TORCH_INTERNAL_ASSERT(
+      tpMessage.payloads.size() == 1,
+      "message expected to contain 1 payload, whereas it contained ",
+      tpMessage.payloads.size(),
+      " payloads");
+  std::vector<char> payload(tpMessage.payloads[0].length);
+  tpMessage.payloads[0].data = (uint8_t*)(payload.data());
```
@lw If we look at this one, the check is just valid for current interface usage. When start to handle duplicated tensor storage and put all tensors' metadata to payload[1], this will break. Does that mean this should be removed..?

lw replied:
This check is being performed _inside_ the function, not by any of its users or by the tests.
